owlkettle/widgets.nim

```diff
@@ -2012,6 +2012,66 @@ renderable ModelButton of BaseWidget:
             proc clicked() =
               echo "Clicked " & $it
 
+renderable SearchEntry of BaseWidget:
+  text: string
+  # child: GtkWidget # This is currently not supported
+  searchDelay: uint = 100 ## Determines the minimum time after a `searchChanged` event occurred before the next can be emitted. Only available when compiling for gtk 4.8
+  placeholderText: string = "Search" ## Only available when compiling for gtk 4.10
+  
+  proc activate(searchString: string) ## Triggered when the user "activated" the search e.g. by hitting "enter" key while SearchEntry is in focus.
+  proc nextMatch() ## Triggered when the user hits the "next entry" keybinding while the search entry is in focus, which is Ctrl-g by default. 
+  proc previousMatch() ## Triggered when the user hits the "previous entry" keybinding while the search entry is in focus, which is Ctrl-Shift-g by default.
+  proc changed(searchString: string) ## Triggered when the user types in the SearchEntry.
+  proc searchStarted(searchString: string)
+  proc stopSearch(searchString: string) ## Triggered when the user "stops" a search, e.g. by hitting the "Esc" key while SearchEntry is in focus. 
+  
+  hooks:
+    beforeBuild:
+      state.internalWidget = gtk_search_entry_new()
+    connectEvents:
+      proc searchCallback(widget: GtkWidget, data: ptr EventObj[proc(searchString: string)]) =
+        let searchString = $gtk_editable_get_text(widget)
+        data[].callback(searchString)
+        data[].redraw()
+      
+      proc changedCallback(widget: GtkWidget, data: ptr EventObj[proc(searchString: string)]) =
```

Comment:
Note: I separated out the changedCallback from the searchCallback.
Only the changedCallback gets to update the State.
I'm not sure if that's okay or not okay, but it seemed right to me as I would always assume the last value that `changed` returns to me to be the value that is written in text.

Reply:
If the last value of `changed` is indeed always the value currently in the `SearchEntry` that is good. Although I would remove the `searchString` parameter from all other events in order to avoid confusion (& force the user to do a proper 2-way binding). Is there anything that prevents you from updating the state in the other events though?

Reply:
Not really and honestly I don't think it makes much of a difference. I can do both, you got a preference (removing search-string from the other callbacks vs. every callback gets to update the `text` field) ?

Reply:
I think after playing with the example it makes more sense to only update in `changed`.

Reply:
Done